### ISSUE TITLE
Some problematic results calculated from libxc

### ISSUE THREAD

**Issue by hoshishin** [pyscf/pyscf#797]
Hi PySCF devs, I used the following density in a specific grid, when using BLYP functional deriv=2 or deriv=1 will give different results.

```
rho_a = numpy.array([ [1.96813585e-11], [-5.27011775e-11],  [1.56257568e-26],  [4.13823869e-14]])
rho_b = numpy.array([ [2.05739056e-11], [-5.50911693e-11],  [4.51859527e-27],  [1.81623010e-12]])

ni.eval_xc('blyp', (rho_a, rho_b), spin=1,relativity=0, deriv=2,verbose=0)[:2]
ni.eval_xc('blyp', (rho_a, rho_b), spin=1,relativity=0, deriv=1,verbose=0)[:2]
```
Exc, Vxc will all be zero if deriv=2. I was puzzled about that. 

Thanks a lot!!!

**Comment by hoshishin** [comment]
Hi PySCF devs, I've read the code in libxc.py, in line 1317 which seems that those functionals may have singularity in those rho conditions. But is this cutoff right to set exc and vxc both 0? And Why xcfun will give non-zero results?

Thanks a lot!

**Comment by sunqm** [comment]
In certain versions of libxc, some high order derivatives of LYP functional are
numerically instable. It gives inf or nan.  libxc.py filters these points. At
these points the value of exc and all high order derivatives (vxc, fxc, kxc) are
set to zero.

xcfun uses different techniques to compute derivatives. I'm not sure whether
xcfun actually produces more accurate results. As it does not break the
calculation, no density or densities are filtered.

On Sat, Jan 16, 2021 at 06:42:14PM -0800, hoshishin wrote:
> Hi PySCF devs, I've read the code in libxc.py, in line 1317 which seems that those functionals may have singularity in those rho conditions. But is this cutoff right to set exc and vxc both 0? And Why xcfun will give non-zero results?
> 
> Thanks a lot!
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/pyscf/pyscf/issues/797#issuecomment-761719901

--

**Comment by hoshishin** [comment]
Thank you a lot. I'm not sure whether both exc and vxc should be set to zeros.

**Comment by gkc1000** [comment]
Generally when the density is 0, you want exc and vxc to be 0, the
analytical forms of the functionals satisfy this constraint.

**Comment by hoshishin** [comment]
Thank you, gkc! Is that mean when rho is small, both xcfun and libxc should set the exc and vxc to 0?

**Comment by gkc1000** [comment]
If they worked in infinite numerical precision, there would be no problem.
Otherwise, the numerical library is typically computing something like
eps1/eps2 * eps3 ... eg. rho^(4/3) * (|grad rho| / rho^(4/3)); in this
latter case clearly the rho^(4/3) factors should cancel, but they will not
precisely cancel if the expression is parenthesized as above, because |grad
rho|/rho^(4/3) diverges as r->infinity for a typically small molecular
density of the form exp(-lambda r).

So the numerical stability is implementation dependent, it is possible to
implement things in a more stable way. PySCF does the practical thing and
simply sets inf/Nan when it sees them to 0. However, I believe there may
still be large numerical errors in xcfun (even if it does not give NaN) for
certain near zero densities and gradients. This what Qiming means by he
does not know if xcfun results are more accurate, they are just not inf or
NaN.

**Comment by hoshishin** [comment]
Thank you, gkc! I see the point. If only some high order derivatives of LYP functional are numerically instable, when I set deriv=2 or deriv=1, what I expect to see is that the exc and vxc calculated from those two parameters should be the same. If deriv=2, exc = 0 and vxc = 0, but for deriv = 1 exc != 0 and vxc != 0, does that mean exc and vxc calculated from deriv = 1 are numerical unstable? Thank you very much!

**Comment by susilehtola** [comment]
You should retry the calculations with the 2.0 prerelease that uses Libxc 5.1.x where the numerical stability of a large number of functionals has undergone serious improvements.

**Comment by hoshishin** [comment]
Thank you, susilehtola! I'll try it.

**Comment by sunqm** [comment]
Fixed in libxc 5.1
